File: frontend/src/types/task.ts
/**
 * 任务状态枚举
 */
export type TaskStatus =
  | 'PENDING'
  | 'QUEUED'
  | 'PROCESSING'
  | 'COMPLETED'
  | 'FAILED'
  | 'ANALYZE_QUEUED'
  | 'ANALYZING'
  | 'ANALYZE_FAILED'
  | 'DETECTING'
  | 'REVIEW_PENDING'
  | 'REVIEW_APPROVED'
  | 'SPLITTING'
  | 'TIMELINE_READY'

export const ACTIVE_TASK_STATUSES: TaskStatus[] = [
  'QUEUED',
  'PROCESSING',
  'ANALYZE_QUEUED',
  'ANALYZING',
  'DETECTING',
  'REVIEW_APPROVED',
  'SPLITTING',
]

export const PROGRESS_VISIBLE_TASK_STATUSES: TaskStatus[] = [
  ...ACTIVE_TASK_STATUSES,
]

export const TERMINAL_TASK_STATUSES: TaskStatus[] = [
  'COMPLETED',
  'FAILED',
  'ANALYZE_FAILED',
  'REVIEW_PENDING',
  'TIMELINE_READY',
]

export const PREVIEWABLE_TASK_STATUSES: TaskStatus[] = [
  'REVIEW_PENDING',
  'REVIEW_APPROVED',
  'SPLITTING',
  'TIMELINE_READY',
  'COMPLETED',
  'FAILED',
  'ANALYZE_FAILED',
]

export const DELETABLE_TASK_STATUSES: TaskStatus[] = [
  'PENDING',
  'QUEUED',
  'PROCESSING',
  'DETECTING',
  'REVIEW_PENDING',
  'REVIEW_APPROVED',
  'SPLITTING',
  'TIMELINE_READY',
  'COMPLETED',
  'FAILED',
  'ANALYZE_QUEUED',
  'ANALYZING',
  'ANALYZE_FAILED',
]

export interface ReviewScene {
  startMs: number
  endMs: number
}

export interface DetectionResult {
  scenes: ReviewScene[]
  durationMs: number
  report: Record<string, unknown>
}

export type DetectorMode = 'content' | 'adaptive'
export type DetectionMode = 'fast' | 'precision'
export type ProcessingMode = 'manual'

export interface ProcessingConfig {
  detector: DetectorMode
  detectionMode: DetectionMode
  useTransnet: boolean
  sceneThreshold: number
  minSceneLenFrames: number
  minSceneDurationMsFloor: number
  downscale: number
  frameSkip: number
  adaptiveThreshold: number
  adaptiveMinContentVal: number
  adaptiveFrameWindow: number
  thumbnailRetry: number
  weightHue: number
  weightSat: number
  weightLum: number
  weightEdges: number
  transnetThreshold: number
  transnetSoftCandidateMultiplier: number
  transnetToleranceFrames: number
  transnetWindowSize: number
  transnetAdditionalBoundaryThreshold: number
  transnetOnlyMinGapFrames: number
  useThresholdDetector: boolean
  thresholdDetectorThreshold: number
  thresholdDetectorFadeBias: number
  mergeGapFrames: number
  splitCopyMode: boolean
  scenedetectTimeoutSec: number
}

export interface ProcessTaskOptions {
  mode?: ProcessingMode
  profile?: 'quality'
  overrideConfig?: Partial<ProcessingConfig>
}

export type ProcessingFieldGroup = 'sensitivity' | 'min-scene' | 'stability' | 'transnet' | 'misc'

export interface ProcessingConfigGroupMeta {
  id: ProcessingFieldGroup
  label: string
  description: string
}

export interface ProcessingConfigFieldMeta {
  key: keyof ProcessingConfig
  type: 'enum' | 'bool' | 'int' | 'float'
  default: string | number | boolean
  min?: number | null
  max?: number | null
  step?: number | null
  group?: ProcessingFieldGroup | null
  label?: string | null
  description?: string | null
  detectorScope?: 'adaptive' | 'content' | null
  modeScope?: 'precision' | null
  boolScope?: keyof ProcessingConfig | null
  options?: string[] | null
  provenance: 'pyscenedetect' | 'smartcut'
  sourceRef?: string | null
  uiVisible: boolean
}

export interface ProcessingConfigMeta {
  version: string
  groups: ProcessingConfigGroupMeta[]
  defaults: ProcessingConfig
  fields: ProcessingConfigFieldMeta[]
}

export interface ProcessTaskResult {
  status: string
  jobId?: string
  deduplicated?: boolean
  resolvedConfig?: Partial<ProcessingConfig>
  configMeta?: ProcessingConfigMeta
}

export interface QualityFlags {
  overSegmented?: boolean
  underSegmented?: boolean
  unstableBoundary?: boolean
  flashFalseCut?: boolean
  shotsPerMinute?: number
  shortShotRatio?: number
  veryShortRatio?: number
  longShotRatio?: number
  sceneCount?: number
  metrics?: Record<string, number>
  [key: string]: unknown
}

export interface SuspectSegment {
  sequenceIndex: number
  startMs: number
  endMs: number
  reason: string
  confidence: number
  reviewHint?: string
}

export interface SplitStats {
  totalScenes: number
  reusedCount: number
  renderedCount: number
  reusedRatio: number
  incrementalEnabled?: boolean
  fallbackFullResplit: boolean
  linkSuccessCount?: number
  copyFallbackCount?: number
  planElapsedMs?: number
  reuseMaterializeElapsedMs?: number
  renderElapsedMs?: number
  dbElapsedMs?: number
  cleanupElapsedMs?: number
  totalElapsedMs?: number
}

/**
 * 任务实体
 */
export interface Task {
  id: string
  displayName: string
  filePath: string
  fileSize: number
  durationMs?: number | null
  status: TaskStatus
  progress: number
  activeOperation?: string | null
  activeJobId?: string | null
  totalScenes: number | null
  shotsCount?: number | null
  previewThumbnailPath?: string | null
  processMode?: string | null
  configProfile?: string | null
  requestedConfig?: Partial<ProcessingConfig> | null
  resolvedConfig?: Partial<ProcessingConfig> | null
  qualityFlags?: QualityFlags | null
  suspectSegments?: SuspectSegment[] | null
  tuningHistory?: Array<Record<string, unknown>> | null
  reviewNotes?: string | null
  latestSplitStats?: SplitStats | null
  detectionResult?: DetectionResult | null
  userEditedScenes?: ReviewScene[] | null
  reviewedAt?: string | null
  createdAt: string
  updatedAt: string
}

/**
 * 镜头/片段实体
 */
export interface Scene {
  id: string
  taskId: string
  sequenceIndex: number
  startMs: number
  endMs: number
  filePath: string | null
  thumbnailPath: string | null
  createdAt: string
}

/**
 * 创建任务 DTO
 */
export interface CreateTaskDTO {
  displayName: string
  file: File
}

/**
 * 上传进度
 */
export interface UploadProgress {
  loaded: number
  total: number
  percentage: number
}

export interface ProgressStreamEvent {
  type: 'progress' | 'terminal'
  taskId: string
  progress: number
  status: TaskStatus
  totalScenes?: number | null
  timestamp: number
}

export interface ProgressStreamErrorEvent {
  type: 'error'
  taskId: string
  errorCode: string
  errorMessage: string
  timestamp: number
}

export type ProgressStreamPayload = ProgressStreamEvent | ProgressStreamErrorEvent

/**
 * 任务详情（包含镜头列表）
 */
export interface TaskDetail extends Task {
  scenes: Scene[]
}

/**
 * API 响应包装
 */
export interface ApiResponse<T> {
  data: T
  message?: string
}

/**
 * 错误响应
 */
export interface ApiError {
  detail: string
  status?: number
}

File: backend/app/schemas/task.py
from datetime import datetime
from typing import Any, Literal, Optional

from pydantic import BaseModel, Field, model_validator


class TaskCreate(BaseModel):
    display_name: str


class ProcessConfigOverride(BaseModel):
    detector: Optional[Literal["content", "adaptive"]] = None
    detection_mode: Optional[Literal["fast", "precision"]] = None
    use_transnet: Optional[bool] = None
    scene_threshold: Optional[float] = None
    min_scene_len_frames: Optional[int] = None
    min_scene_duration_ms_floor: Optional[int] = None
    downscale: Optional[int] = None
    frame_skip: Optional[int] = None
    adaptive_threshold: Optional[float] = None
    adaptive_min_content_val: Optional[float] = None
    adaptive_frame_window: Optional[int] = None
    thumbnail_retry: Optional[int] = None
    weight_hue: Optional[float] = None
    weight_sat: Optional[float] = None
    weight_lum: Optional[float] = None
    weight_edges: Optional[float] = None
    transnet_threshold: Optional[float] = None
    transnet_soft_candidate_multiplier: Optional[float] = None
    transnet_tolerance_frames: Optional[int] = None
    transnet_window_size: Optional[int] = None
    transnet_additional_boundary_threshold: Optional[float] = None
    transnet_only_min_gap_frames: Optional[int] = None
    use_threshold_detector: Optional[bool] = None
    threshold_detector_threshold: Optional[float] = None
    threshold_detector_fade_bias: Optional[float] = None
    merge_gap_frames: Optional[int] = None
    split_copy_mode: Optional[bool] = None
    scenedetect_timeout_sec: Optional[int] = None


class ProcessTaskRequest(BaseModel):
    mode: Literal["manual"] = "manual"
    profile: Literal["quality"] = "quality"
    override_config: Optional[ProcessConfigOverride] = None


class ReviewScenePayload(BaseModel):
    start_ms: int = Field(ge=0)
    end_ms: int = Field(gt=0)

    @model_validator(mode="after")
    def validate_range(self):
        if self.end_ms <= self.start_ms:
            raise ValueError("end_ms must be greater than start_ms")
        return self


class SaveReviewDataRequest(BaseModel):
    scenes: list[ReviewScenePayload]


class LocalPrecisionPreviewRequest(BaseModel):
    anchor_scene_index: int = Field(ge=0)
    radius: int = Field(default=2, ge=0, le=10)


class LocalPrecisionTargetRange(BaseModel):
    start_ms: int = Field(ge=0)
    end_ms: int = Field(gt=0)
    start_index: int = Field(ge=0)
    end_index: int = Field(ge=0)

    @model_validator(mode="after")
    def validate_range(self):
        if self.end_ms <= self.start_ms:
            raise ValueError("end_ms must be greater than start_ms")
        if self.end_index < self.start_index:
            raise ValueError("end_index must be greater than or equal to start_index")
        return self


class LocalPrecisionProposal(BaseModel):
    target_range: LocalPrecisionTargetRange
    original_scenes: list[ReviewScenePayload]
    proposed_scenes: list[ReviewScenePayload]
    report: dict[str, Any]


class LocalPrecisionPreviewResponse(LocalPrecisionProposal):
    pass


class TaskResponse(BaseModel):
    id: str
    display_name: str
    file_path: str
    file_size: int
    duration_ms: Optional[int] = None
    status: str
    progress: int
    active_operation: Optional[str] = None
    active_job_id: Optional[str] = None
    total_scenes: Optional[int]
    shots_count: Optional[int] = None
    preview_thumbnail_path: Optional[str] = None
    process_mode: Optional[str] = None
    config_profile: Optional[str] = None
    requested_config: Optional[dict[str, Any]] = None
    resolved_config: Optional[dict[str, Any]] = None
    quality_flags: Optional[dict[str, Any]] = None
    suspect_segments: Optional[list[dict[str, Any]]] = None
    tuning_history: Optional[list[dict[str, Any]]] = None
    review_notes: Optional[str] = None
    latest_split_stats: Optional[dict[str, Any]] = None
    detection_result: Optional[dict[str, Any]] = None
    user_edited_scenes: Optional[list[dict[str, Any]]] = None
    reviewed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class SceneResponse(BaseModel):
    id: str
    task_id: str
    sequence_index: int
    start_ms: int
    end_ms: int
    file_path: Optional[str]
    thumbnail_path: Optional[str]
    created_at: datetime

    class Config:
        from_attributes = True


class ProcessTaskResponse(BaseModel):
    status: str
    job_id: Optional[str] = None
    deduplicated: bool = False
    resolved_config: Optional[dict[str, Any]] = None
    config_meta: Optional[dict[str, Any]] = None
